src/pytorch_lightning/core/module.py

```diff
@@ -172,6 +172,23 @@ def lr_schedulers(self) -> Optional[Union[LRSchedulerTypeUnion, List[LRScheduler
         # multiple schedulers
         return lr_schedulers
 
+    @property
+    def trainer(self) -> "pl.Trainer":
+        if self._trainer is None:
+            raise RuntimeError(f"{self.__class__.__qualname__} is not attached to a `Trainer`.")
+        return self._trainer
+
+    @trainer.setter
+    def trainer(self, trainer: "pl.Trainer") -> None:
+        if not isinstance(trainer, pl.Trainer):
```

Comment:
```suggestion
        if not isinstance(trainer, pl.Trainer) and trainer is not None:
```
this would allow us to also recursively detach the trainer again in case we want to copy the model, pickle it (e.g. want to send it to another process)

Reply:
@justusschock True. In that case, I'm leaning towards removing this check entirely and letting duck-typing happen.

What do you think? And should the Loop.trainer property have the same change?

Reply:
I still think it's good to have some check and not allow anything there. One more question: Do you think we should maybe make this a weakref? So that we don't actually have circular references?

For the Loop.trainer:
Is there ever a way where we do need to copy those or detach? From what I can think of, we don't meaning we are fine with just setting it, right?

Reply:
> Do you think we should maybe make this a weakref?

This will block pickling or deepcopying the LM, but it seems like that's already an issue considering the existence of the `_should_prevent_trainer_and_dataloaders_deepcopy` flag:

https://github.com/Lightning-AI/lightning/blob/feb8e7d344dae74d95dcacba95eaf75c2893cf3a/src/pytorch_lightning/core/module.py#L1966-L1967

Let me try setting the weakref to see if any tests complain.

> Is there ever a way where we do need to copy those or detach? From what I can think of, we don't meaning we are fine with just setting it, right?

Correct